### ISSUE TITLE
Organise > Column: Reshape adding another append

### ISSUE THREAD

**Issue by rdstern** [africanmathsinitiative/R-Instat#1902]
We currently have a general dialogue for merging data from 2 data frames.

And we can "Append Data Frames".  This append is for multiple data frames and is what could be called vertical, i.e. if we have 3 data frames with 6 columns and 36 rows in each, then the resulting data frame would have 6 columns (if all the names match) and there will be 36*3=108 rows of data.  I think this uses the cbind command or an equivalent.  The restriction is that (at least some of the) column names should match.  If not, then the data are still appended, but there are many missing values where the names don't match.  

(In fact, if no names at all match, then it might be good to check for that, and ask whether the user wishes to continue?)

The final dialogue we need is to be able to append data frames horizontally.  I think this will use rbind (or an equivalent).  In the above example it would produce a new data frame with 36 rows and 18 columns.

The restriction now is that the data frames need all to be the same length.  I guess rbind might complain otherwise and we might check for that situation.  So we only proceed if they are the same length.  The message could suggest the full merge dialogue be used.

There are 2 ways we could do this - either adding the option vertical/horizontal to the existing append dialogue.  Or we could have a different dialogue.  I think I favour a new dialogue.  Then the exiting dialogue might now be called "Append Rows" and the new Dialogue would be "Append Columns", or "Join Columns".  (I think of "Append" as downwards, so slightly prefer "Join Columns".

In the menu now put a line under Unstack.  Then have Merge and these 2 dialogues and then another line.

The new dialogue is really simple - now, because it is the same as the current Append one, except that we don't need the checkbox for the ID column.  The new data frame name could be "Join".

By the way, what happens to these names if you do this more than once?  Does the code detect that the name Append (or Join) already exists and so it checnges that, e.g. Join1, etc?

A simple example to test the new dialogue could use the library dataset called Anscombe.  That gives 8 short columns, called x1, x2, x3, x4.  And y1, y2, y3, y4.  Now stack the x's into a new data frame.  And then stack the y's.  Now we want to join these 2 data frames.

Roger

**Comment by dannyparsons** [comment]
The command is `cbind` for appending columns (we current use `rbind.fill` I think to do the Append Rows). The standard `cbind` from base will be good enough for us, there is no fast equivalent in `dplyr` etc. as there is for rows.

I think I favour a separate dialog too, as they are quite different operations.

I would call it "Append Columns" as we are using the term join already in the merge dialog (left join etc. and "joining columns") and also the merge R functions are actually called `left_join`, `full_join` etc. so I think avoiding join is less confusing here. Append Rows and Append Columns seems clear enough for the dialog names.

In terms of the data frame name, it should be able to detect current names and use append1 etc. as the default. We still don't have warnings when you replace an existing data frame though.

**Comment by Lunalo** [comment]
I have been testing the commands cbind,cbind2,cbind.data.frame and same for the rows
cbind in general can bind many columns of matrix or dataframes of the same number of rows, cbind2 binds two columns of the same length. Now, cbind.data.frame does exactly what cbind does but convert the output to data frame in case it is not. Same applies to the rbind, rbind2, rbind.data.frame. 

I will give more thoughts shortly..

Thanks.

**Comment by dannyparsons** [comment]
which package is `cbind2` in?

**Comment by Lunalo** [comment]
it is **methods** package

**Comment by dannyparsons** [comment]
`cbind2` seems to be only for S4 objects, which we don't need. `cbind.data.frame` is the specific method `cbind` uses when any of the arguments are a data frame. Our arguments should always be data frames so `cbind` will do the same as `cbind.data.frame` for us.
There is also `bind_cols` from `dplyr`. We had issues using the equivalent, `bind_rows`, in the Instat Object so I would be nervous to use `bind_cols` but it could be tested. An advantage is that `bind_cols` should be much faster than `cbind`. So `cbind` or `bind_cols` would be my suggestions.

For `rbind` there are more alternatives, including `rbind.fill` from `plyr`. We looked into that in detail before because we needed to test how each deals with attributes, so I think we have the functions we want here but feel free to look into that too.
